src/types.rs

```diff
@@ -586,6 +589,106 @@ impl ZeroSsl {
     }
 }
 
+/// A unique certificate identifier for the ACME renewal information (ARI) extension
+///
+/// See <https://www.ietf.org/archive/id/draft-ietf-acme-ari-07.html#section-4.1> for
+/// more information.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CertificateIdentifier<'a> {
+    /// The BASE64URL-encoded authority key identifier (AKI) extension `keyIdentifier` of the certificate
+    pub authority_key_identifier: Cow<'a, str>,
+
+    /// The BASE64URL-encoded serial number of the certificate
+    pub serial: Cow<'a, str>,
+}
+
+impl CertificateIdentifier<'_> {
+    /// Encode a unique certificate identifier using the provided authority key ID and serial
+    ///
+    /// `authority_key_identifier` must be the DER-encoded ASN.1 octet string from the
+    /// `keyIdentifier` field of the `AuthorityKeyIdentifier` extension found in the certificate
+    /// to be identified.
+    ///
+    /// `serial` must be the DER-encoded ASN.1 serial number from the certificate to be identified.
+    /// Care must be taken to use the **encoded** serial number, not a big integer representation.
+    ///
+    /// The combination uniquely identifies a certificate within all certificates issued by the
+    /// same CA.
+    ///
+    /// See [RFC 5280 §4.1.2.2], [RFC 5280 §4.2.1.1], and [draft-ietf-acme-ari-07 §4.1]
+    ///
+    /// [RFC 5280 §4.1.2.2]: https://www.rfc-editor.org/rfc/rfc5280#section-4.1.2.2
+    /// [RFC 5280 §4.2.1.1]: https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.1
+    /// [draft-ietf-acme-ari-07 §4.1]: https://www.ietf.org/archive/id/draft-ietf-acme-ari-07.html#section-4.1
+    pub fn new(authority_key_identifier: Der<'_>, serial: Der<'_>) -> Self {
+        Self {
+            authority_key_identifier: BASE64_URL_SAFE_NO_PAD
+                .encode(authority_key_identifier)
+                .into(),
+            serial: BASE64_URL_SAFE_NO_PAD.encode(serial).into(),
+        }
+    }
+}
+
+impl<'de: 'a, 'a> Deserialize<'de> for CertificateIdentifier<'a> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct CertificateIdentifierVisitor<'de> {
+            marker: PhantomData<CertificateIdentifier<'de>>,
+        }
```

Comment:
One of my pet peeves is that custom `Deserialize` implementations are very verbose. I've found that a good approach to improve that is to define an intermediate type and implement the `Deserialize` implementation in terms of a processing step on top of a derived `Deserialize` implementation. Sort of like this:

```rust
#[derive(Deserialize)]
struct CertificateIdentifierStr<'a>(&'a str);

let s = CertificateIdentifierStr::deserialize(deserializer)?;
let Some((aki, serial)) = value.split_once('.') else {
     ..
};

if serial.contains('.') {
     return Err(..);
}

Ok(CertificateIdentifier { .. })
```

Reply:
Ohhh that's so much nicer. Thank you. I think I've only had to do this ~2 or so times and didn't think about this approach.

Reworked!